cluster-autoscaler/main.go

```diff
@@ -684,6 +692,15 @@ func main() {
 	featureGate.AddFlag(pflag.CommandLine)
 	kube_flag.InitFlags()
 
+	// If the DRA flag is passed, we need to set the DRA feature gate as well. The selection of scheduler plugins for the default
+	// scheduling profile depends on feature gates, and the DRA plugin is only included if the DRA feature gate is enabled. The DRA
+	// plugin itself also checks the DRA feature gate and doesn't do anything if it's not enabled.
+	if *enableDynamicResourceAllocation && !featureGate.Enabled(features.DynamicResourceAllocation) {
```

Comment:
I don't want to slow down integration but it's a bit of a shame we can't simply wire in the k/k feature gate interface in here for the user to specify natively in CAS (rather than manually setting it here based on CAS runtime config)

Reply:
> CA already allows you to pass the feature gate flag

Do you mean inherit a feature gate from kubelet, kube-scheduler, etc, or use the feature gate facility instead of the command line config facility? I would prefer using a feature gate approach (user expresses a feature gate config to the CAS runtime) so that we could eventually graduate it (first set to true by default and then retire altogether like k/k does) and simplify the code here. I wouldn't want to have "if DRA then do this" code paths here forever.

Reply:
> Do you mean inherit a feature gate from kubelet, kube-scheduler, etc, or use the feature gate facility instead of the command line config facility?

I don't think the feature gate flags are "centralized" or "inherited" between different components, it seems that they are set separately for each component (see the second sentence here: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#overview). The flags are likely synchronized between the components by something that deploys Kubernetes, but I'm not sure if that something would know to pass the flag to Cluster Autoscaler anyway.

I fully agree with not keeping the DRA flag forever in CA and removing it once we're confident in the implementation, but I think we can do that with a CA-specific flag as well. `--enable-dynamic-resource-allocation` is basically a lightweight, CA-specific feature gate - we've been following this pattern for previous features as well.

Especially because I'm pretty sure the graduation criteria are not the same here - DRA went to beta in 1.32, but DRA autoscaling is definitely an alpha in 1.32. If we were to just follow the graduation path for the DRA feature gate, we'd IMO be graduating things to early in CA. It seems that we'd have more control over the graduation if we use a dedicated flag.

Reply:
Sounds like we can do this for the CA-specific `--enable-dynamic-resource-allocation` flag:

1) set default to true when DRA goes GA in k/k (for example, 1.34 or 1.35)
2) deprecate flag w/ a 4 release warning ("Warning, this flag will go away in 1.38")
3) remove flag altogether in (e.g.,) 1.38

Those are just example k8s versions, but something like that will allow us to simplify a lot of this code in the next few years, which will be beneficial to the project.

Reply:
Yes, this is definitely the plan!

Reply:
(hopefully earlier than 1.38 :sweat_smile:)